yarpctest/recorder/recorder.go

```diff
@@ -0,0 +1,378 @@
+// remote service. Every request and its response is recorded into a YAML file,
+// under the directory "testdata/recordings".
+//
+// During replay, the client doesn't need to be connected, for any recorded
+// request Recorder will return the recorded response. Any new request will
+// return an error.
+type Recorder struct {
+	mode       Mode
+	logger     TestingT
+	recordsDir string
+}
+
+const defaultRecorderDir = "testdata/recordings"
+
+// Mode is the recording mode of the recorder.
+type Mode int
+
+const (
+	// invalidMode is private and used to represent invalid modes.
+	invalidMode = iota
+
+	// Replay replays stored request/response pairs, any non pre-recorded
+	// requests will be rejected.
+	Replay
+
+	// Overwrite will store all request/response pairs, overwriting existing
+	// records.
+	Overwrite
+
+	// Append will store all new request/response pairs and replay from
+	// existing record.
+	Append
+)
+
+func (m Mode) toHumanString() string {
+	switch m {
+	case Replay:
+		return "replaying"
+	case Overwrite:
+		return "recording (overwrite)"
+	case Append:
+		return "recording (append)"
+	default:
+		return fmt.Sprintf("Mode(%d)", int(m))
+	}
+}
+
+// ModeFromString converts an English string of a mode to a `Mode`.
+func ModeFromString(s string) (Mode, error) {
+	switch s {
+	case "replay":
+		return Replay, nil
+	case "overwrite":
+		return Overwrite, nil
+	case "append":
+		return Append, nil
+	}
+	return invalidMode, fmt.Errorf(`invalid mode: "%s"`, s)
+}
+
+// TestingT is an interface used by the recorder for logging and reporting fatal
+// errors. It is intentionally made to match with testing.T.
+type TestingT interface {
+	// Logf must behaves similarly to testing.T.Logf.
+	Logf(format string, args ...interface{})
+
+	// Fatal should behaves similarly to testing.T.Fatal. Namely, it must abort
+	// the current test.
+	Fatal(args ...interface{})
+}
+
+// NewRecorder returns a Recorder in whatever mode specified via the
+// `--recorder` flag.
+//
+// The new Recorder instance is a yarpc filter middleware. It takes a logger as
+// argument compatible with `testing.T`.
+//
+// See package documentation for more details.
+func NewRecorder(logger TestingT, opts ...Option) *Recorder {
+	cwd, err := os.Getwd()
+	if err != nil {
+		logger.Fatal(err)
+	}
+	recorder := &Recorder{
+		mode:       invalidMode,
+		logger:     logger,
+		recordsDir: filepath.Join(cwd, defaultRecorderDir),
+	}
+	var mode Mode = invalidMode
+	for _, opt := range opts {
+		if opt.RecordsPath != "" {
+			recorder.recordsDir = opt.RecordsPath
+		}
+		if opt.Mode != invalidMode {
+			mode = opt.Mode
+		}
+	}
+	if mode == invalidMode {
+		mode, err = ModeFromString(*recorderFlag)
+		if err != nil {
+			logger.Fatal(err)
+		}
+	}
+	recorder.SetMode(mode)
+	return recorder
+}
+
+// Option can be passed to NewRecorder for customization.
+type Option struct {
+	Mode        Mode   // Initial mode.
+	RecordsPath string // Initial records location.
+}
```

Comment:
I'm confused about this. You have a single struct with all the options but you're still accepting a vararg of them? Why? Is it so that the API is usable without passing in a second parameter or were you trying to use the options pattern? For the options pattern?